rectools/models/nn/transformers/lightning.py

```diff
@@ -0,0 +1,407 @@
+        return loss
+
+    def _xavier_normal_init(self) -> None:
+        for _, param in self.torch_model.named_parameters():
+            if param.data.dim() > 1:
+                torch.nn.init.xavier_normal_(param.data)
+
+    def _prepare_for_inference(self, recommend_device: tp.Optional[str]) -> None:
+        if recommend_device is None:
+            recommend_device = "cuda" if torch.cuda.is_available() else "cpu"
+        device = torch.device(recommend_device)
+        self.torch_model.to(device)
+        self.torch_model.eval()
+
+    def _get_user_item_embeddings(
+        self,
+        recommend_dataloader: DataLoader,
+        recommend_device: tp.Optional[str],
+    ) -> tp.Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Prepare user embeddings for all user interaction sequences in `recommend_dataloader`.
+        Prepare item embeddings for full items catalog.
+        """
+        self._prepare_for_inference(recommend_device)
+        device = self.torch_model.item_model.device
+
+        with torch.no_grad():
+            item_embs = self.torch_model.item_model.get_all_embeddings()
+            user_embs = []
+            for batch in recommend_dataloader:
+                batch_embs = self.torch_model.encode_sessions(batch["x"].to(device), item_embs)[:, -1, :]
+                user_embs.append(batch_embs)
+
+        return torch.cat(user_embs), item_embs
+
+    def _recommend_u2i(
+        self,
+        user_ids: InternalIdsArray,
+        recommend_dataloader: DataLoader,
+        sorted_item_ids_to_recommend: InternalIdsArray,
+        k: int,
+        dataset: Dataset,  # [n_rec_users x n_items + n_item_extra_tokens]
+        filter_viewed: bool,
+        recommend_use_torch_ranking: bool,
+        recommend_n_threads: int,
+        recommend_device: tp.Optional[str],
+    ) -> InternalRecoTriplet:
+        """Recommend to users."""
+        ui_csr_for_filter = None
+        if filter_viewed:
+            ui_csr_for_filter = dataset.get_user_item_matrix(include_weights=False, include_warm_items=True)[user_ids]
+
+        user_embs, item_embs = self._get_user_item_embeddings(recommend_dataloader, recommend_device)
+
+        ranker: Ranker
+        if recommend_use_torch_ranking:
+            ranker = TorchRanker(
+                distance=Distance.DOT,
+                device=item_embs.device,
+                subjects_factors=user_embs[user_ids],
+                objects_factors=item_embs,
+            )
+        else:
+            user_embs_np = user_embs.detach().cpu().numpy()
+            item_embs_np = item_embs.detach().cpu().numpy()
+            del user_embs, item_embs
+            torch.cuda.empty_cache()
+            ranker = ImplicitRanker(
+                Distance.DOT,
+                user_embs_np[user_ids],  # [n_rec_users, n_factors]
+                item_embs_np,  # [n_items + n_item_extra_tokens, n_factors]
+                num_threads=recommend_n_threads,
+                use_gpu=False,
```

Comment:
Why is it so?  
I mean, I understand the logic  
But it's unclear that if I put `recommend_use_torch_ranking = False`, then GPU won't be used as well
Maybe we can add an error or at least a warning here that if we were asked to use implicit and device is not None and not CPU, then sorry, we will use cpu

Reply:
Removed ImplicitRanker